js-library/src/services/ens.ts

```diff
@@ -0,0 +1,171 @@
+
+/**
+ * ENS Class
+ * Provides interface for interaction with the ENS smart contracts
+ */
+export class ENS {
+  private _provider: providers.JsonRpcProvider
+  private _ensRegistryContract: Contract
+  private _subdomainRegistrarContract: Contract
+  private _publicResolverContract: Contract
+
+  constructor(
+    environment: Environment = Environment.LOCALHOST,
+    signerOrProvider: SignerOrProvider | null = null,
+    private domain = ENS_DOMAIN,
+  ) {
+    this._provider = new providers.JsonRpcProvider(ENVIRONMENT_RPC_URLS[environment])
+
+    const { ENSRegistryContract, PublicResolverContract, SubdomainRegistrarContract } =
+      CONTRACTS_METADATA[environment]
+
+    this._ensRegistryContract = new Contract(
+      ENS_REGISTRY_ADDRESS as string,
+      ENSRegistryContract.abi,
+      this._provider,
+    )
+    this._publicResolverContract = new Contract(
+      PUBLIC_RESOLVER_ADDRESS as string,
+      PublicResolverContract.abi,
+      this._provider,
+    )
+    this._subdomainRegistrarContract = new Contract(
+      SUBDOMAIN_REGISTRAR_ADDRESS as string,
+      SubdomainRegistrarContract.abi,
+      this._provider,
+    )
+
+    if (signerOrProvider) {
+      this.connect(signerOrProvider)
+    }
+  }
+
+  /**
+   * returns RPC provider
+   */
+  public get provider(): providers.JsonRpcProvider {
+    return this._provider
+  }
+
+  /**
+   * Connects signer to the smart contracts
+   * @param signerOrProvider An instance of ethers.js Wallet or any other signer
+   */
+  public connect(signerOrProvider: SignerOrProvider) {
+    this._publicResolverContract = this._publicResolverContract.connect(signerOrProvider)
+    this._subdomainRegistrarContract = this._subdomainRegistrarContract.connect(signerOrProvider)
+    this._ensRegistryContract = this._ensRegistryContract.connect(signerOrProvider)
+  }
+
+  /**
+   * Returns owner address of the provided username on ENS
+   * @param username ENS username
+   * @returns owner's address
+   */
+  public async getUsernameOwner(username: string): Promise<EthAddress> {
+    const usernameHash = this.hashUsername(username)
+
+    return this._ensRegistryContract.owner(usernameHash)
+  }
+
+  /**
+   * Checks whether username is available on ENS
+   * @param username ENS username
+   * @returns True if the username is available
+   */
+  public async isUsernameAvailable(username: string): Promise<boolean> {
+    const owner = await this.getUsernameOwner(username)
+    return owner === NULL_ADDRESS
+  }
+
+  /**
+   * Sets owner of the provided username on ENS
+   * @param username ENS username
+   * @param address Owner address of the username
+   */
+  public async registerUsername(username: string, address: EthAddress, publicKey: PublicKey): Promise<void> {
+    const ownerAddress = await this.getUsernameOwner(username)
+
+    if (ownerAddress !== NULL_ADDRESS && ownerAddress !== address) {
+      throw new Error(`ENS: Username ${username} is not available`)
+    }
```

Comment:
Yes, the contract method will throw an error if the username has already been registered. But there is no way to inform the user about the error. Because, whatever the reason why transaction is rejected, ganache  returns this error: `cannot estimate gas; transaction may fail or may require manual gas limit`. With this implementation the user can know the reason why it failed.